samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -267,22 +271,34 @@ public TestRunner addOutputStream(CollectionStream stream) {
     return this;
   }
 
+
   /**
    * Utility to run a test configured using TestRunner
+   *
+   * @param timeout time to wait for the high level application or low level task to finish. This timeout does not
+   *                include input stream initialization time or the assertion time over output streams.
+   *                This timeout just accounts for time that samza job takes run
    */
-  public void run() {
+  public void run(Duration timeout) throws SamzaException, TimeoutException {
     Preconditions.checkState((app == null && taskClass != null) || (app != null && taskClass == null),
         "TestRunner should run for Low Level Task api or High Level Application Api");
     final LocalApplicationRunner runner = new LocalApplicationRunner(new MapConfig(configs));
+    boolean timedOut = false;
     if (app == null) {
       runner.runTask();
-      runner.waitForFinish();
+      timedOut = !runner.waitForFinish(timeout);
     } else {
       runner.run(app);
-      runner.waitForFinish();
+      timedOut = !runner.waitForFinish(timeout);
+    }
+    if (timedOut) {
+      throw new TimeoutException("Timed out waiting for application to finish");
+    }
+    ApplicationStatus status = runner.status(app);
+    if (status.getStatusCode() == ApplicationStatus.StatusCode.UnsuccessfulFinish) {
+      throw new SamzaException(status.getThrowable());
```

Comment:
We will not get the complete stacktrace by creating `SamzaException` from throwable object.

Resultant `SamzaException` will contain the head of exception message. We can do either of the following:

1. Just bubble up the throwable: `throw status.getThrowable()`.
2. Use `ExceptionUtils`(from `apache-commons`) to get complete stacktrace from the throwable: `throw new SamzaException(ExceptionUtils.getStackTrace(t))`

Reply:
Used ExceptionUtils from apache-commons to get the stack trace!